### DISCUSSION TITLE
Isn't there "task.cancel()"?

### DISCUSSION THREAD

**Discussion by kubotafamily** [pybricks/support#1948] [Q&A] [answered]
We will be participating in the FLL Challenge.
I created an awaitable function for each run, but I am unable to execute this function when a button is pressed, and terminate this function (but not terminate the program) when the button is pressed again.
Is there something like asyncio's task.cancel()?

**Comment by laurensvalk** [comment, marked as answer]
Here's how you can do that, in both blocks and Python:

![image](https://github.com/user-attachments/assets/25c388d1-3eec-4bb3-a70f-edeadfa04aa0)

**Reply by kubotafamily** [reply]
Thank you very much!

I have one more question.
``` Python
from pybricks.tools import wait, hub_menu, multitask, run_task

async def subtask1():
    await foo1()
    while True:
        bar1()
        await wait(1)

async def subtask2():
    await foo2()
    while True:
        bar2()
        await wait(1)

...

async def until_pressed() -> None:
    while Button.CENTER not in hub.buttons.pressed():
        await wait(1)

runs = [subtask1, subtask2, ...]
run_num = 0

while True:
    selected = hub_menu(*[(i + run_num) % len(runs) + 1 for i in range(0, len(runs))])
    run_task(multitask(runs[selected - 1](), until_pressed(), race=True))
    run_num = selected - 1

```
I have written a program like this, but if the button is pressed while "foo1()" or "foo2()" is running, I want "run_num" to remain the same, and if the button is pressed while "bar1()" or "bar2()" is running, I want "run_num" to be incremented by 1.
How can I implement this?

Also, even though I have written "hub.system.set_stop_button(Button.BLUETOOTH)", the program does not end when I press the Bluetooth button. I would appreciate some help.

Thank you in advance.

**Reply by laurensvalk** [reply]
> I have written a program like this, but if the button is pressed while "foo1()" or "foo2()" is running, I want "run_num" to remain the same, and if the button is pressed while "bar1()" or "bar2()" is running, I want "run_num" to be incremented by 1.
How can I implement this?

Perhaps before functions, choose


```
foo_busy = False
```

Then change your sub functions as:

```python
async def subtask1():
    global foo_busy
    foo_busy = True
    await foo1()
    foo_busy = False
    while True:
        bar1()
        await wait(1)
```

Then later in your main loop you can choose whether or not to increment the counter by checking if `foo_busy` is `True` or `False`.

> Also, even though I have written "hub.system.set_stop_button(Button.BLUETOOTH)", the program does not end when I press the Bluetooth button. I would appreciate some help.

Do you have a small program that reproduces this? Then we can try it out and fix it if there is an issue.

**Reply by kubotafamily** [reply]
submerged.py
``` Python
from pybricks.hubs import PrimeHub
from pybricks.parameters import Axis, Button
from pybricks.tools import wait, hub_menu, multitask, run_task

hub = PrimeHub(Axis.Z, -Axis.Y)
hub.system.set_stop_button(Button.BLUETOOTH)

foo_busy = False
async def until_pressed() -> None:
    while Button.CENTER not in hub.buttons.pressed():
        await wait(1)

```

runFLL.py
``` Python
from submerged import *

async def subtask1():
    global foo_busy
    await foo1()
    foo_busy = False
    while True:
        bar1()
        await wait(1)

async def subtask2():
    global foo_busy
    await foo2()
    foo_busy = False
    while True:
        bar2()
        await wait(1)

...

runs = [subtask1, subtask2, ...]
run_num = 0

while True:
    selected = hub_menu(*[(i + run_num) % len(runs) + 1 for i in range(0, len(runs))])
    foo_busy = True
    run_task(multitask(runs[selected - 1](), until_pressed(), race=True))
    if foo_busy:
        foo_busy = False
        run_num = selected - 1
    else:
        run_num = selected

```

I am running "runFLL.py" with a program like this, but the program does not terminate when I press the Bluetooth button.
Why is that?

**Reply by laurensvalk** [reply]
It would be hard to run any code with missing pieces. Can you perhaps create a separate (but self contained) program that reproduces the issue? Thank you!

**Reply by kubotafamily** [reply]
I finally figured out when this problem occurs.
Pressing the Bluetooth button while "hub_menu()" is running doesn't stop the program, but it does stop when "subtask1()" or "subtask2()" are running.
Here's a program that reproduces the problem.

``` Python
from pybricks.hubs import PrimeHub
from pybricks.parameters import Axis, Button
from pybricks.tools import wait, hub_menu, multitask, run_task

hub = PrimeHub()
hub.system.set_stop_button(Button.BLUETOOTH)

foo_busy = False
async def until_pressed():
    while Button.CENTER not in hub.buttons.pressed():
        await wait(1)

async def subtask1():
    global foo_busy
    await wait(5000)
    await hub.speaker.beep()
    foo_busy = False
    while True:
        await wait(1)

async def subtask2():
    global foo_busy
    await wait(5000)
    await hub.speaker.beep()
    foo_busy = False
    while True:
        await wait(1)

runs = [subtask1, subtask2]
run_num = 0

while True:
    selected = hub_menu(*[(i + run_num) % len(runs) + 1 for i in range(0, len(runs))])
    foo_busy = True
    hub.speaker.beep()
    run_task(multitask(runs[selected - 1](), until_pressed(), race=True))
    if foo_busy:
        foo_busy = False
        run_num = selected - 1
    else:
        run_num = selected

```
